File: keycloak/26.7.1/analyzed/model/jpa/src/main/java/org/keycloak/models/jpa/PaginationUtils.java
/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.keycloak.models.jpa;

import jakarta.persistence.TypedQuery;

/**
 * JPA 查询分页工具：统一处理 first/max 参数及 Hibernate 分页边界问题。
 */
public class PaginationUtils {

    /** 未指定 max 时的默认上限，规避 HHH-14295 分页缺陷。 */
    public static final int DEFAULT_MAX_RESULTS = Integer.MAX_VALUE >> 1;

    /** 对 TypedQuery 应用 offset/limit；first 有效且 max 缺失时使用 {@link #DEFAULT_MAX_RESULTS}。 */
    public static <T> TypedQuery<T> paginateQuery(TypedQuery<T> query, Integer first, Integer max) {
        if (first != null && first >= 0) {
            query = query.setFirstResult(first);

            // Hibernate 在 setFirstResult 后必须 setMaxResults，否则可能抛出异常
            if (max == null || max < 0) {
                max = DEFAULT_MAX_RESULTS;
            }
        }

        if (max != null && max >= 0) {
            query = query.setMaxResults(max);
        }

        return query;
    }

}

File: ragflow/0.26.4/analyzed/api/utils/pagination_utils.py
#
#  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
"""
REST API 分页约束：page_size 不得超过公开上限。
"""

#

# REST 公开 API 单页最大条数
REST_API_MAX_PAGE_SIZE = 100


def validate_rest_api_page_size(page_size: int) -> int:
    # 校验 page_size ≤ REST_API_MAX_PAGE_SIZE
    """Validate REST API page_size values against the public maximum."""
    if page_size > REST_API_MAX_PAGE_SIZE:
        raise ValueError(f"page_size must be less than or equal to {REST_API_MAX_PAGE_SIZE}")
    return page_size
